####                                      MEET PATEL
####                                      MISY 408

## Project 7- MISY408

In this project you should examine Which Factors are significant in predicting Insurance **charges**.

The following are the attributes of the dataset:

**age:** age of primary beneficiary

**sex:** gender of primary beneficiary

**bmi:** body mass index of primary beneficiary

**children:** Number of children covered by health insurance / Number of dependents.

**smoker:** whther the person is smoker or non-smoker

**region:** The beneficiary's residental area in the USA.

**charges:** the cost of health insurance

**Question 1:** Import Required libraries, load the dataset to df dataframe, display first 5 rows.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('insurance.csv')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

**Question 2:** 

* Show the shape of dataset
* Check for missing values and if there is any missing values drop them.
* Drop duplicate rows.
* Show the statistical summary of numerical variables (describe)

In [2]:
df.shape

(1338, 7)

In [3]:
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [4]:
df.drop_duplicates(inplace=True)

In [5]:
df.describe().transpose()

count          mean           std        min       25%        50%  \
age       1337.0     39.222139     14.044333    18.0000    27.000    39.0000   
bmi       1337.0     30.663452      6.100468    15.9600    26.290    30.4000   
children  1337.0      1.095737      1.205571     0.0000     0.000     1.0000   
charges   1337.0  13279.121487  12110.359656  1121.8739  4746.344  9386.1613   

                  75%          max  
age          51.00000     64.00000  
bmi          34.70000     53.13000  
children      2.00000      5.00000  
charges   16657.71745  63770.42801

**Question 3:Preliminary analysis** 

a. Create a correlation matrix of the variables in the df dataframe. 

In [6]:
df.corr()

age       bmi  children   charges
age       1.000000  0.109344  0.041536  0.298308
bmi       0.109344  1.000000  0.012755  0.198401
children  0.041536  0.012755  1.000000  0.067389
charges   0.298308  0.198401  0.067389  1.000000

b. Do some Preliminary analysis that can help you with selecting the best independent variables for predicting dependent variable (suggestion: for example you can compare average charges for differnt smoker groups, sex, region, and etc.)

In [7]:
df.groupby('region')['charges'].mean()

region
northeast    13406.384516
northwest    12450.840844
southeast    14735.411438
southwest    12346.937377
Name: charges, dtype: float64

In [8]:
df.groupby('smoker')['charges'].mean()

smoker
no      8440.660307
yes    32050.231832
Name: charges, dtype: float64

b. Interpert your correlation matrix and preliminary analysis. 
For example: Based on the correlation matrix and preliminary analysis which features can be good predictors for dependent variable (charges)? 

Based on the correlation matrix above, we can see that all the variables have significant and positive correlation amongst one and another. Variables such as AGE and BMI (num) would be a good indicator for the predictor output (Charges). Also, the smoker variable would be a good indicator for the predictor variable (Charges) as it will indicate the number of charges a specific beneficiary will receive. 

**Question 4**

Use get_dummies() method to encode the values of categorical columns. Display the first 5 rows of df dataframe.

In [9]:
df2 = df[['age', 'bmi', 'smoker', 'charges']].copy()
df2.head()

age     bmi smoker      charges
0   19  27.900    yes  16884.92400
1   18  33.770     no   1725.55230
2   28  33.000     no   4449.46200
3   33  22.705     no  21984.47061
4   32  28.880     no   3866.85520

In [10]:
df2 = pd.get_dummies(df2)
df2.head()

age     bmi      charges  smoker_no  smoker_yes
0   19  27.900  16884.92400          0           1
1   18  33.770   1725.55230          1           0
2   28  33.000   4449.46200          1           0
3   33  22.705  21984.47061          1           0
4   32  28.880   3866.85520          1           0

**Question 5**

a. Specify the X and Y dataframes (X is independent variables. Y is dependent variable ('charges'))

Note: You should choose independent variables yourself (you can use the correlation matrix and preliminary data analysis results as a guide for choosing independent variables)

In [18]:
X = df2[['age', 'bmi','smoker_no']].copy()
Y = df2['charges'].copy()

a. Split the data to train and test set (30 percent of data should be test set)

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.3)

**Question 6**

Run OLS Regression model on train set

In [20]:
import statsmodels.api as sm
import math
from sklearn.metrics import mean_squared_error

X_train = sm.add_constant(X_train)
est = sm.OLS(y_train,X_train).fit()
predictions = est.predict()
print(est.summary())
print("\nAverage error: {:.2f}.".format(math.sqrt(est.mse_resid)))

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.749
Model:                            OLS   Adj. R-squared:                  0.749
Method:                 Least Squares   F-statistic:                     928.4
Date:                Sun, 28 Nov 2021   Prob (F-statistic):          3.08e-279
Time:                        23:42:09   Log-Likelihood:                -9452.2
No. Observations:                 935   AIC:                         1.891e+04
Df Residuals:                     931   BIC:                         1.893e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.115e+04   1146.292      9.723      0.0

b. Interpret the results (R-Squared, Prob (F-statistic), coefficients of independent variables, p-values of independent variables, Average Error). Check Adj. R-Squared to make sure there is no model overfit.

Based on the table above, we can see that the R-squared is 74.9% - which would indicate that change in charges can be represented by the independent variable. The F-statistic Prob is less than .05 which would suggest tha the model is linear.

age - The coefficient for the age variable would be 265.72. Therefore as age increase the charges will refelect an increase in 250.63 dollars. 

bmi = The coefficient for the Bmi variable is 288.82.Therefore as BMI increases the chargesw ill reflect an increase in 349.55 in charge price. 

Smoker_no: The coefficient for the non Smoker is -24590. Therefore would indicate a negative correlation and so as there are fewer smokers the charges would decreasse by -23480

The p value for the following variables is less than .05, which would indenfiy these variables as significant. 

**Question 7**

a. Run OLS Regression model on test set

In [23]:
import statsmodels.api as sm

X_test = sm.add_constant(X_test) 
est = sm.OLS(y_test,X_test).fit()
predictions = est.predict()
print(est.summary())
print("\nAverage error: {:.2f}.".format(math.sqrt(est.mse_resid)))

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.746
Method:                 Least Squares   F-statistic:                     393.6
Date:                Sun, 28 Nov 2021   Prob (F-statistic):          1.14e-118
Time:                        23:42:26   Log-Likelihood:                -4089.0
No. Observations:                 402   AIC:                             8186.
Df Residuals:                     398   BIC:                             8202.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.471e+04   1913.922      7.683      0.0

b.Interpret the results (R-Squared, Prob (F-statistic), coefficients of independent variables, p-values of independent variables, Average Error). Check Adj. R-Squared to make sure there is no model overfit.

The R-squared is 74.8%, which would indicate that the change in charges can be justified by the independent variables. The model respresents an overfit. The F-statistic Prob is linear.

age- The coefficent is 276.57.83. Therefore as age increases the charge will increase by 256.83. 

BMI - The coefficient is 258.63.64. As BMI increase, the charges increase by 259.64

Smoker_no - The coefficient is -24690. as the number of non-smokers increase the charges decrease by -24340

The p-value for the all the variables is less than .05. 

c. Check whether your regression model on Test set has Multicollinearity problem

In [15]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

variables = est.model.exog 
vif = pd.DataFrame() 
vif["VIF Factor"] = [variance_inflation_factor(variables, i) for i in range(variables.shape[1])]
vif["features"] = X_test.columns
print('VIF: {}'.format(vif))

VIF:    VIF Factor   features
0   34.107955      const
1    1.017843        age
2    1.017737        bmi
3    1.000286  smoker_no


d. Explain why your model has Multicollinearity problem or does not have Multicollinearity problem. Which independent variables have Multicollinearity problem. If you have multicolinearity problem, what is your approach for solving Multicollinearity problem?

There is no multicolinearity problem with this regression model.Therefore because the VIF factor for all specified variables is under 2. Therefore, there is no need to examine any further. 

e. Compare your Regression model on Test set and Train set. Is there model overfit? Please explain why there is model overfit or why there is not model overfit?

After conducting a regression model on the Test set and Train set there is no significant indication of a model overfit. Therefore, becasue all of the variables are signficant and also because the  change in coefficents between variables between test and train set for bmi and age is not signficant enough to indicate that there is a model overfit. 

**Question 8 (Answer this question if your model needs improvement)**

If your model needs to be improved ( for example if a new variable should be added to the model or a variable should be dropped from the model), In the following run a new regression on train set and test set.

a. Specify the X and Y dataframes (X is independent variables. Y is dependent variable)

b. Run OLS Regression model on train set and interpret the results

c. Run OLS Regression model on test set and interpret the results